flight/PiOS/Common/pios_servo.c

```diff
@@ -101,188 +97,152 @@ int32_t PIOS_Servo_Init(const struct pios_servo_cfg * cfg)
 		return -1;
 	}
 	memset(output_channel_mode, 0, servo_cfg->num_channels * sizeof(typeof(output_channel_mode)));
-#endif
+
 
 	return 0;
 }
 
 /**
  * @brief PIOS_Servo_SetMode Sets the PWM output frequency and resolution.
- * A speed of 0 indicates Synchronous updates (e.g. OneShot), in which case
- * the period is set to the maximal value. Otherwise the prescalar will be
- * determined by the PWM mode to set the resolution, and the period will be
- * calculated based on the speed. The information required to convert from us
- * to compare value is cached for each channel (not timer) to facilitate 
- * PIOS_Servo_Set and PIOS_SERVO_HPWM_Set.
- * @param speeds array of rates in Hz
- * @param pwm_mode mode of the timer (1us to 1/12us)
+ * An output rate of 0 indicates Synchronous updates (e.g. SyncPWM/OneShot), otherwise
+ * normal PWM at the specified output rate. SyncPWM uses hardware one-pulse mode.
+ * Timer prescaler and related parameters are calculated based on the channel
+ * with highest max pulse length on each timer. A deadtime is provided to
+ * ensure SyncPWM pulses cannot merge.
+ * The information required to convert from us to compare value is cached
+ * for each channel (not timer) to facilitate PIOS_Servo_Set
+ * @param out_rate array of output rate in Hz, banks elements
  * @param banks maximum number of banks
+ * @param channel_max array of max pulse lengths, number of channels elements
  */
-void PIOS_Servo_SetMode(const uint16_t * speeds, const enum pwm_mode *pwm_mode, uint8_t banks)
+void PIOS_Servo_SetMode(const uint16_t *out_rate, const int banks, const uint16_t *channel_max)
 {
-	if (!servo_cfg) {
+	if (!servo_cfg || banks > PIOS_SERVO_MAX_BANKS)
 		return;
+
+	struct timer_bank {
+		TIM_TypeDef *timer;
+		uint32_t clk_rate;
+		uint16_t max_pulse;
+		uint16_t prescaler;
+		uint16_t period;
+	} timer_banks[PIOS_SERVO_MAX_BANKS];
+
+	memset(&timer_banks, 0, sizeof(timer_banks));
+	int bank = 0;
+
+	// find max pulse length for each bank and channel allocation
+	for (int i = 0; i < servo_cfg->num_channels && bank < banks; i++) {
+		bool new = i > 0;
+		const struct pios_tim_channel *chan = &servo_cfg->channels[i];
+		/* See if any previous channels use that same timer */
+		for (int j = 0; j < i && new; j++) {
+			if (chan->timer == servo_cfg->channels[j].timer)
+				new = false;
+		}
+		if (new)
+			++bank;
+
+		timer_banks[bank].timer = chan->timer;
+		timer_banks[bank].max_pulse = MAX(timer_banks[bank].max_pulse, channel_max[i]);
 	}
 
+	// configure timers/banks
 	TIM_TimeBaseInitTypeDef TIM_TimeBaseStructure = servo_cfg->tim_base_init;
 	TIM_TimeBaseStructure.TIM_ClockDivision = TIM_CKD_DIV1;
 	TIM_TimeBaseStructure.TIM_CounterMode = TIM_CounterMode_Up;
+	for (int i = 0; i < banks; i++) {
+		/* Calculate the maximum clock frequency for the timer */
+		volatile uint32_t max_tim_clock = max_timer_clock(timer_banks[i].timer);
+		// check if valid timer
+		if (!max_tim_clock)
+			return;
```

Comment:
Maybe assert would be appropriate. The existing code used return for all fatal cases.